sdks/python/apache_beam/runners/portability/portable_runner.py

```diff
@@ -95,31 +95,40 @@ def default_docker_image():
   @staticmethod
   def _create_environment(options):
     portable_options = options.view_as(PortableOptions)
+    return PortableRunner._resolve_environment(
+        portable_options.environment_type,
+        portable_options.environment_config)
+
+  @staticmethod
+  def _resolve_environment(environment_type, environment_config):
     environment_urn = common_urns.environments.DOCKER.urn
-    if portable_options.environment_type == 'DOCKER':
+    if environment_type == 'DOCKER':
       environment_urn = common_urns.environments.DOCKER.urn
-    elif portable_options.environment_type == 'PROCESS':
+    elif environment_type == 'PROCESS':
       environment_urn = common_urns.environments.PROCESS.urn
-    elif portable_options.environment_type in ('EXTERNAL', 'LOOPBACK'):
+    elif environment_type in ('EXTERNAL', 'LOOPBACK'):
       environment_urn = common_urns.environments.EXTERNAL.urn
-    elif portable_options.environment_type:
-      if portable_options.environment_type.startswith('beam:env:'):
-        environment_urn = portable_options.environment_type
+    elif environment_type == 'EMBEDDED':
+      environment_urn = environment_type
+    elif environment_type:
+      # Allow embedded environment for testing
+      if environment_type.startswith('beam:env:') or \
+            environment_type == 'EMBEDDED':
```

Comment:
Here we seem to be embedding knowledge about Java's understanding about the special string "EMBEDDED" meaning "java embedded." There's also a python "embedded" that is separate from this. https://github.com/apache/beam/blob/release-2.13.0/sdks/python/apache_beam/portability/python_urns.py#L38 On the other hand, specifying "DOCKER" would give the Python docker image, right? Seems we should be doing something more principled here.

Reply:
No, the environment_type is a URN. `Embedded` is a special testing URN that I didn't want to add as a standard urn. So this alongside with the payload works across all SDKs.

Reply:
It doesn't have to be a standard URN, but if we're spreading it across multiple codebases we should probably qualify it because it has a specific meaning. Maybe "beam:env:embedded_java_for_testing:v1" or something like that then. It'd break in strange ways if I tried to use this with a python expansion service, or on a non-Java runner. What you really want to say is "run this as cheaply as possible."

(Also, I still think it's inconsistent to have a "EMBEDDED" mean a java-runner-specific thing thing and "DOCKER" a python-image-spawning one and "PROCESS" something else altogether.)

Reply:
Agree, we should give it a proper, private URN. Yes, reusing the existing Python code here has some problematic implications. We probably only want to accept the URN, or come up with a better way to specify that we are using the embedded environment of an SDK (e.g. `Embedded:Java`)